evm/src/cpu/dup_swap.rs

```diff
@@ -0,0 +1,249 @@
+    used_in: usize,
+    used_out: usize,
+    filter: P,
+    lv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    for i in used_in..NUM_GP_CHANNELS - used_out {
+        yield_constr.constraint(filter * lv.mem_channels[i].used);
+    }
+}
+
+/// Disable unused channels (everything except the first `used_in` and the last `used_out`).
+pub fn disable_unused_channels_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    used_in: usize,
+    used_out: usize,
+    filter: ExtensionTarget<D>,
+    lv: &CpuColumnsView<ExtensionTarget<D>>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    for i in used_in..NUM_GP_CHANNELS - used_out {
+        let constr = builder.mul_extension(filter, lv.mem_channels[i].used);
+        yield_constr.constraint(builder, constr);
+    }
+}
+
+pub fn eval_packed_dup<P: PackedField>(
+    n: P,
+    lv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let filter = lv.is_cpu_cycle * lv.op.dup;
+
+    let in_channel = &lv.mem_channels[0];
+    let out_channel = &lv.mem_channels[NUM_GP_CHANNELS - 1];
+
+    channels_equal_packed(filter, in_channel, out_channel, yield_constr);
+
+    constrain_channel_packed(true, filter, n, in_channel, lv, yield_constr);
+    // `out_channel` is constrained by the usual stack constraints.
+
+    disable_unused_channels_packed(1, 1, filter, lv, yield_constr);
+}
+
+pub fn eval_ext_circuit_dup<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    n: ExtensionTarget<D>,
+    lv: &CpuColumnsView<ExtensionTarget<D>>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    let filter = builder.mul_extension(lv.is_cpu_cycle, lv.op.dup);
+
+    let in_channel = &lv.mem_channels[0];
+    let out_channel = &lv.mem_channels[NUM_GP_CHANNELS - 1];
+
+    channels_equal_ext_circuit(builder, filter, in_channel, out_channel, yield_constr);
+
+    constrain_channel_ext_circuit(builder, true, filter, n, in_channel, lv, yield_constr);
+    // `out_channel` is constrained by the usual stack constraints.
+
+    disable_unused_channels_ext_circuit(builder, 1, 1, filter, lv, yield_constr);
+}
+
+pub fn eval_packed_swap<P: PackedField>(
+    n: P,
+    lv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let filter = lv.is_cpu_cycle * lv.op.swap;
+
+    let in1_channel = &lv.mem_channels[0];
+    let in2_channel = &lv.mem_channels[1];
+    let out1_channel = &lv.mem_channels[NUM_GP_CHANNELS - 2];
+    let out2_channel = &lv.mem_channels[NUM_GP_CHANNELS - 1];
+
+    channels_equal_packed(filter, in1_channel, out1_channel, yield_constr);
+    channels_equal_packed(filter, in2_channel, out2_channel, yield_constr);
+
+    constrain_channel_packed(true, filter, n, in2_channel, lv, yield_constr);
+    constrain_channel_packed(false, filter, n, out1_channel, lv, yield_constr);
```

Comment:
This feels slightly confusing to me since half the logic is here and half in stack.rs; takes a bit of thinking to piece it together and check that each step occurs in the right order / with the right offsets.

How would you feel about using `num_pops: false`, `pushes: false` for SWAP so that the logic can be centralized here, with 4 `constrain_channel_packed` calls?

Reply:
Yeah, fair point!